feat: subgroups [1/3]

GamerGirlandCo opened the pull request:
this is the first in a series of 3 pull requests to implement subgroups which closes #1872
### checklist
- [x] models
  - [x] add `GroupID` field to `Repository` struct
  - [x] new (to add)
    - [x] `Group`
    - [x] group-team relation
      stores a team’s [access level](https://github.com/go-gitea/gitea/blob/main/models%2Fperm%2Faccess_mode.go#L14) to a group
    - [x] `GroupUnit`
      like a repository unit, but grants or denies access to a specific unit at the group level. will override unit permissions defined on a [team](https://github.com/go-gitea/gitea/blob/main/models%2Forganization%2Fteam_unit.go)
- [x] misc
	- [x] update [GetUserRepoPermission](https://github.com/go-gitea/gitea/blob/main/models/perm/access/repo_permission.go#L193) function to take into account group permissions
	- [x] update [GetTeamRepositories](https://github.com/go-gitea/gitea/blob/main/models/repo/org_repo.go#L30) to take into account group permissions
	- [x] update condition builder functions in [models/repo/repo_list.go](https://github.com/go-gitea/gitea/blob/main/models/repo/repo_list.go)  to check for group access
- [x] API
    access to these endpoints is governed by `read:organization` and `write:organization` scopes
  - [x] POST `/org/{org}/groups/new` → create a new group at the root level of an org
  - [x] `/groups/{group_id}`
    - [x] GET → retrieves the group
    - [x] PATCH → updates the group's name and/or description
  - [x] GET `/groups/{group_id}/subgroups` → get a group's subgroups
  - [x] GET `/groups/{group_id}/repos` → get a group's repos
  - [x] POST `/groups/{group_id}/new` → creates a new subgroup inside a group
  - [x] POST `/groups/{group_id}/move` → moves a group to different group in the same organization, or to the root level if `newParent` is null or 0
  - [x] POST `/{username}/{reponame}/groups/move` → moves a repository to a different group within an organization, or to the root level if `newParent` is null or 0
  - [x] allow a group ID to be specified when creating a repository
---
<sub>
/claim #1872
</sub>

<!-- branch-stack-start -->

-------------------------
- main
  - **feat: subgroups [1/3]** :point_left:
    - #37630

<sup>[Stack](https://www.git-town.com/how-to/proposal-breadcrumb.html) generated by [Git Town](https://github.com/git-town/git-town)</sup>

<!-- branch-stack-end -->

lunny commented:
It seems reusing `organization` as a sub `organization` could reduce many duplicated works.

GamerGirlandCo committed:
feat: add group model

GamerGirlandCo committed:
feat: create GroupList type and methods

GamerGirlandCo committed:
refactor: add avatar to group

GamerGirlandCo committed:
refactor: add models/group/errors.go

GamerGirlandCo committed:
chore: update makefile and add new tool

this tool will generate group-related swagger schemas without repeating the same comments twice in different places

GamerGirlandCo committed:
refactor: add unit tests to make sure group reordering works properly

GamerGirlandCo committed:
refactor: update repo model

add `GroupID` and `GroupSortOrder` fields

GamerGirlandCo committed:
refactor: add group-related url segments to list of reserved usernames

GamerGirlandCo committed:
refactor: add `GroupID` to `SearchRepoOptions`

GamerGirlandCo committed:
refactor: update repository storage layout as per https://github.com/go-gitea/gitea/issues/1872#issuecomment-3194681583

GamerGirlandCo committed:
fix: add missing group id parameters to test function calls

GamerGirlandCo committed:
refactor: update integration tests to use new signature of `GetRepositoryByOwnerAndName`

GamerGirlandCo committed:
refactor: update more calls to `GetRepositoryByName`

GamerGirlandCo committed:
refactor: update activity actions to return paths/links with a repo's group

GamerGirlandCo committed:
refactor: update avatar utils to handle group avatars

GamerGirlandCo committed:
refactor: update slice utils

- add `SliceMap` util

works like javascripts's `Array.prototoype.map`, taking in a slice and transforming each element with the provided function

- add dedicated utilities for removing from and inserting into slices

GamerGirlandCo committed:
feat: add API types for groups

GamerGirlandCo committed:
refactor: update `IssueMeta` and `Repository` API structs

- add `GroupID` and `GroupSortOrder` fields to `Repository` struct
- add GroupID field to IssueMeta struct

GamerGirlandCo committed:
feat: add api routes and functions for repository groups

GamerGirlandCo committed:
fix: ensure artifact download endpoint urls are correct

GamerGirlandCo committed:
refactor: update `routers/api/v1/repo/pull.go`

update `GetPullRequestByBaseHead` to parse group IDs out of base:head string

GamerGirlandCo committed:
feat: update `serv` command and hooks to recognize group ids in urls

GamerGirlandCo committed:
feat: add group ID to repository pushing env

GamerGirlandCo committed:
feat: add `models/shared/group` package

this mitigates circular import errors when writing functions referencing both groups and repositories

GamerGirlandCo committed:
fix: ensure queries with CTE don't cause panics during tests

GamerGirlandCo committed:
feat: update private/hook routes to support repository groups

GamerGirlandCo committed:
refactor: add group id to file comparison urls

GamerGirlandCo committed:
refactor: add group id to `runnersCtx` struct

GamerGirlandCo committed:
refactor: update `services/context` package

- add `RepoGroup` struct containing information about the visited group (if any)
- add `RepoGroup` field to both web and api contexts

GamerGirlandCo committed:
feat: add conversion functions for repository groups

GamerGirlandCo committed:
refactor: update repository API response conversion to include group ID and sort order

GamerGirlandCo committed:
feat: add group service

functions included so far:
- avatar uploading/deletion
- group deletion
- group creation
- group moving (including moving item inside a group)
- group update
- group searching (only used in frontend/web components for now)

GamerGirlandCo committed:
refactor: add `GroupID` field to LFS request context struct

GamerGirlandCo committed:
fix: ensure `OwnerName` field is updated in groups owned by an org when its name is updated

GamerGirlandCo committed:
feat: add support for group IDs in issue references

GamerGirlandCo committed:
feat: update functions in the `modules/markup` package to support group IDs

GamerGirlandCo committed:
feat: update base:head comparison string parsing to support references to repos inside groups

GamerGirlandCo committed:
refactor: add LFS routes for repos in subgroups

GamerGirlandCo committed:
feat: update repository URL parsing to support subgroups

GamerGirlandCo committed:
refactor: add `DedupeBy` function to container module

this removes duplicate items from a slice using a custom function

GamerGirlandCo committed:
refactor: add swagger API request and response types

GamerGirlandCo committed:
refactor: add group ID parameter to `CheckCreateRepository` func

GamerGirlandCo committed:
refactor: add missing `GroupID` field whn creating certain structs

GamerGirlandCo committed:
feat: add support for group IDs in issue references

GamerGirlandCo committed:
feat: update `getUniqueRepositoryName` func to take an additional group ID parameter

GamerGirlandCo committed:
refactor: update `repo_permission.go`

ensure that repos in private/limited groups can't be seen without the doer being a member of the owner org

GamerGirlandCo committed:
refactor: modify repo test fixtures to add `group_id` field

GamerGirlandCo committed:
chore: update swagger definitions

GamerGirlandCo committed:
feat: add integration tests

GamerGirlandCo committed:
refactor: update `AccessibleGroupCondition`

- remove unused parameters
- re-add the ability for site admins to be able to see all groups regardless of visibility

GamerGirlandCo committed:
chore: update swagger definitions

GamerGirlandCo committed:
fix: build errors

GamerGirlandCo committed:
style: fix lint errors

GamerGirlandCo committed:
fix: faulty condition in group search

jamilahmadzai committed:
Fix subgroup move ordering guards

Signed-off-by: jamilahmadzai <jamilurrehman722@gmail.com>

GamerGirlandCo committed:
refactor: add `ChildGroupCond` func, which matches a group and all its descendants recursively

GamerGirlandCo committed:
refactor: address various code quality issues

GamerGirlandCo committed:
refactor: update group.go

- update cycle detection to check descendants instead of ancestors
- update `ChildGroupCond` to take a condition as a parameter
- update `ChildGroupCond` to return IDs in ascending order (to enable binary search)

GamerGirlandCo committed:
fix: panic during tests upon encountering multiline CTE

GamerGirlandCo committed:
fix: update cycle detection

ensure we're searching for the new parent in descendants, not the group itself

YongDo-Hyun reviewed (commented):
LGTM.

GamerGirlandCo committed:
refactor: make URL parsing more robust and add new tests

GamerGirlandCo committed:
refactor: address various nits and potential bugs

GamerGirlandCo committed:
refactor: update package imports

GamerGirlandCo committed:
refactor: address linting nits

GamerGirlandCo committed:
fix: failing mysql test

GamerGirlandCo committed:
refactor: update `GetRepoAbsoluteLink` method to support groups

luca-schlecker commented:
I've wondered whether groups are intended to be limited to just organizations.

There's only the http endpoint for orgs (`POST /org/{org}/groups/new`), but the implementation seemed like supporting _any_ owner and thus also personal accounts.

```go
type Group struct {
	// ...
	OwnerID     int64 `xorm:"INDEX NOT NULL"`
	OwnerName   string
	Owner       *user_model.User    `xorm:"-"`
	// ...
}
```

YongDo-Hyun commented:
> I've wondered whether groups are intended to be limited to just organizations.
> 
> There's only the http endpoint for orgs (`POST /org/{org}/groups/new`), but the implementation seemed like supporting _any_ owner and thus also personal accounts.
> 
> ```go
> type Group struct {
> 	// ...
> 	OwnerID     int64 `xorm:"INDEX NOT NULL"`
> 	OwnerName   string
> 	Owner       *user_model.User    `xorm:"-"`
> 	// ...
> }
> ```

However, branching out in personal accounts, like subgroups, is illogical and shouldn't happen.

luca-schlecker commented:
> > I've wondered whether groups are intended to be limited to just organizations.
> > There's only the http endpoint for orgs (`POST /org/{org}/groups/new`), but the implementation seemed like supporting _any_ owner and thus also personal accounts.
> > ```go
> > type Group struct {
> > 	// ...
> > 	OwnerID     int64 `xorm:"INDEX NOT NULL"`
> > 	OwnerName   string
> > 	Owner       *user_model.User    `xorm:"-"`
> > 	// ...
> > }
> > ```
> 
> However, branching out in personal accounts, like subgroups, is illogical and shouldn't happen.

Could you please elaborate on why it is illogical?

I can see subgroups for personal accounts be used for grouping multiple repositories belonging to a shared topic or project. This could significantly decrease clutter. Hierarchical, directory-like structures can be easier to handle than flat lists.

For example, I have multiple projects/topics (like university) that each have multiple repositories. I currently prefix each repository but I'd prefer being able to create a subgroup for them. To me, that is also distinctly different to creating an organization for them.

GamerGirlandCo committed:
refactor: update branch comparison URL funcs

YongDo-Hyun commented:
> > > I've wondered whether groups are intended to be limited to just organizations.
> > > There's only the http endpoint for orgs (`POST /org/{org}/groups/new`), but the implementation seemed like supporting _any_ owner and thus also personal accounts.
> > > ```go
> > > type Group struct {
> > > 	// ...
> > > 	OwnerID     int64 `xorm:"INDEX NOT NULL"`
> > > 	OwnerName   string
> > > 	Owner       *user_model.User    `xorm:"-"`
> > > 	// ...
> > > }
> > > ```
> > 
> > 
> > However, branching out in personal accounts, like subgroups, is illogical and shouldn't happen.
> 
> Could you please elaborate on why it is illogical?
> 
> I can see subgroups for personal accounts be used for grouping multiple repositories belonging to a shared topic or project. This could significantly decrease clutter. Hierarchical, directory-like structures can be easier to handle than flat lists.
> 
> For example, I have multiple projects/topics (like university) that each have multiple repositories. I currently prefix each repository but I'd prefer being able to create a subgroup for them. To me, that is also distinctly different to creating an organization for them.

However, I think it can be handled by creating a group. Because a personal account is personal. It contains your forks for development. For example, to develop your university project, you fork that group/subgroup/againsubgroup/endlesssubgroup/repository, develop an unstable project, try to stabilize it, test it with support from certain CI/CD tools, finalize it, and finally merge it by creating a PR. The main repository is sacred to me, and no matter who you are, unstable code should NOT be added to the main repository. Therefore, it still seems illogical to me.

luca-schlecker commented:
> However, I think it can be handled by creating a group. Because a personal account is personal. It contains your forks for development. For example, to develop your university project, you fork that group/subgroup/againsubgroup/endlesssubgroup/repository, develop an unstable project, try to stabilize it, test it with support from certain CI/CD tools, finalize it, and finally merge it by creating a PR. The main repository is sacred to me, and no matter who you are, unstable code should NOT be added to the main repository. Therefore, it still seems illogical to me.

Ok, I now absolute see where you are coming from and it makes a lot of sense. Yet, this currently seems more based on one's own conventions regarding workflow and thus personal preference.

I'd just leave it up to @GamerGirlandCo and the maintainers to decide where they would like to take this. I just noticed it and thought I'd point it out. 😁

I could also get along with creating a personal organization in which I'd then store everything. Just seemed nicer if this would be an option allowed for personal accounts too.

GamerGirlandCo commented:
> > However, I think it can be handled by creating a group. Because a personal account is personal. It contains your forks for development. For example, to develop your university project, you fork that group/subgroup/againsubgroup/endlesssubgroup/repository, develop an unstable project, try to stabilize it, test it with support from certain CI/CD tools, finalize it, and finally merge it by creating a PR. The main repository is sacred to me, and no matter who you are, unstable code should NOT be added to the main repository. Therefore, it still seems illogical to me.
> 
> Ok, I now absolute see where you are coming from and it makes a lot of sense. Yet, this currently seems more based on one's own conventions regarding workflow and thus personal preference.
> 
> I'd just leave it up to @GamerGirlandCo and the maintainers to decide where they would like to take this. I just noticed it and thought I'd point it out. 😁
> 
> I could also get along with creating a personal organization in which I'd then store everything. Just seemed nicer if this would be an option allowed for personal accounts too.

the reason i restricted this feature to orgs only is because it was also supposed to integrate ACLS via teams, which don't exist on personal accounts. however, implementing ACLS was taking too long, so i decided to cut it out of this branch and maybe add it back in a future PR. maybe i could allow personal accounts to have subgroups too, since i removed the parts blocking it...

mrmelon54 commented:
> maybe i could allow personal accounts to have subgroups too, since i removed the parts blocking it...

I would appreciate being able to make groups under my personal account to tidy up my repositories. The previous comment about only using a personal account for forks assumes that I don't have any personal projects at all, which seems more illogical than not.

GamerGirlandCo committed:
refactor: allow subgroups to be created in personal accounts

GamerGirlandCo committed:
refactor: add ability to filter heatmap activity by group ID

GamerGirlandCo committed:
fix: incorrect path parameters in new API routes

GamerGirlandCo committed:
fix: unit tests failing to compile

GamerGirlandCo committed:
refactor: add `UserGroupLink` method to `Group` struct

GamerGirlandCo committed:
refactor: update `groupAssignment` func

assign the return values of `OrgGroupLink` or `UserGroupLink` methods  to `RepoGroup.OrgGroupLink` depending on whether or not the owner is an org

GamerGirlandCo committed:
refactor: inject more helper values into template context via `GroupAssignmentWeb`

GamerGirlandCo committed:
fix: panic caused by nil `ContextUser`

GamerGirlandCo committed:
fix: ensure that activity outside the current group doesn't show up in dashboard heatmap

GamerGirlandCo committed:
refactor: update migration index before rebase

GamerGirlandCo committed:
refactor: add `GroupID` to `UsesRef` struct

GamerGirlandCo committed:
fix: failing migrations

GamerGirlandCo committed:
refactor: update group service

fix bug that allows groups to be moved to parents belonging to different owners

GamerGirlandCo committed:
fix: update group service

ensure that renaming repos only happens when the group ID is different from the new parent

GamerGirlandCo committed:
refactor: update group assignment

rename `RenderedDescription` -> `RenderedGroupDescription` to avoid conflicts with user bio

GamerGirlandCo committed:
refactor: update `prepareUserProfileTabData` func

ensure we only return root-level repos

GamerGirlandCo committed:
fix: typo in user group creation API url

GamerGirlandCo committed:
refactor: break subgroup cycle detection out into its own function

GamerGirlandCo committed:
refactor: update gorup service

add ownership checks to `NewGroup` func

GamerGirlandCo committed:
style: run formatter

GamerGirlandCo committed:
chore: regenerate swagger templates

GamerGirlandCo committed:
fix: ensure repos in private groups are visible to repo-level collaborators

GamerGirlandCo committed:
refactor: update services/context/group.go

break up `if` statement in group assignment to make it more readable

GamerGirlandCo committed:
fix: build errors after rebase

bircni commented:
_Disclaimer: I just wanted to get a quick overview - here is what was found_

**Branch:** `feat/subgroups/base`  
**Scope:** ~141 files (excluding ~52k lines of generated swagger), core subgroups feature  
**Reviewed with AI**

---

## 1. Structural regressions (presumptive blockers)

### Route registration is duplicated instead of normalized

The biggest architectural smell in this branch is how group-aware repo URLs are wired: almost every repo route is registered twice.

```go
m.Group("/{username}/{reponame}", rootRepoFn, ...)
m.Group("/{username}/group/{group_id}/{reponame}", rootRepoFn, ...)
```

This pattern appears **23 times** in `routers/web/web.go` and **9 times** in `routers/api/v1/api.go`. You started extracting handler closures (`rootRepoFn`, `settingsFn`, etc.), which helps locally, but the real problem remains: **the routing model itself is duplicated across the entire surface area**.

**Code-judo move:** normalize once, route once.

- Add a single registration helper, e.g. `registerRepoRoutes(m, fn, middlewares...)`, that mounts both path shapes internally.
- Or better: early middleware that parses `/{owner}/group/{gid}/{repo}` vs `/{owner}/{repo}` into a canonical `RepoPath` on context, then keep **one** route tree.

Right now every future repo route requires a human to remember the duplicate. That is guaranteed spaghetti growth.

`routers/web/web.go` was already **1,779 lines** at base; this branch pushes it to **1,831**. The file was already far past a healthy boundary. Adding more mirrored routes makes a bad situation worse instead of decomposing routing first.

---

### `AccessibleParentGroupCond` runs a live query while building a SQL condition

```403:423:models/group/group.go
func AccessibleParentGroupCond(ctx context.Context, idStr string, user *user_model.User) builder.Cond {
	accessibleCond := AccessibleGroupCondition(user)
	unionSQL := groupHierarchyCTEBuilder(
		accessibleCond.And(builder.Eq{
			"parent_group_id": 0,
		}))
	// ...
	err := e.SQL(sql + " select id from group_hierarchy").Find(&ids)
	if err != nil {
		return builder.NotIn(idStr)
	}
	return builder.In(idStr, ids)
}
```

This is called from `IsPrivateBecauseOfParentPermissions`, which is invoked on **every org repo permission resolution** when a group exists:

```469:477:models/perm/access/repo_permission.go
	if group != nil {
		isGroupPrivate = group.Visibility == structs.VisibleTypePrivate
		isPrivateBecauseOfParent, err := group.IsPrivateBecauseOfParentPermissions(ctx, user)
		// ...
		isGroupPrivate = isGroupPrivate || isPrivateBecauseOfParent
	}
```

Problems:

1. **Side effects during condition construction** — a `builder.Cond` factory should not hit the database.
2. **Over-fetching** — it materializes *all* accessible root groups and descendants, then checks one group ID.
3. **Silent failure** — DB error becomes `NOT IN (...)`, which is effectively “deny everything.”
4. **Hot-path cost** — this will run constantly (repo pages, API, git/http, search).

**Code-judo move:** replace this with a pure SQL subquery/CTE that answers “is group X visible via ancestor chain for user Y?” for a **single** group ID. `ChildGroupCond` already shows the CTE style; `ParentGroupCond` should use the same family of primitives, not a separate N+1 walker + a separate global CTE.

---

### Non-recursive `DeleteGroup` reparents repos in DB but not on disk

`MoveGroupItem` correctly renames git storage when `group_id` changes:

```243:260:services/group/group.go
			if repo.GroupID != opts.NewParent {
				ndir := filepath.Dir(...)
				// MkdirAll + gitrepo.RenameRepository(...)
			}
			if err = MoveRepositoryToGroup(ctx, repo, opts.NewParent, opts.NewPos); err != nil {
```

But `DeleteGroup` (non-recursive) only updates DB rows:

```86:93:services/group/delete.go
		for i, repo := range repos {
			repo.GroupID = toDelete.ParentGroupID
			repo.GroupSortOrder = int(inParent + int64(i))
			if _, err = sess.Table(&repo_model.Repository{}).ID(repo.ID).Update(repo); err != nil {
```

Storage paths use `owner/{groupID}/repo.git` when grouped. Reparenting without `RenameRepository` leaves DB and filesystem diverged. This looks like a real correctness bug, and it exists because delete/move logic was not centralized.

**Fix shape:** non-recursive delete should call the same move primitive (`MoveGroupItem` / `MoveRepositoryToGroup` + rename), not bespoke SQL updates.

---

## 2. Missed dramatic simplifications

### Three parallel permission implementations for the same concept

Group authorization is implemented in at least three places with overlapping rules:

| Layer | Location |
|-------|----------|
| Model | `Group.CanAccess*`, `IsOwnedBy`, `IsMemberOf`, `IsAdminOf` |
| Web/API context | `groupAssignment` in `services/context/group.go` |
| API middleware | `reqGroupMembership` in `routers/api/v1/api.go` |

`reqGroupMembership` re-derives org owner/admin/member, then calls `CanAccessAtLevel`, then re-checks `DoerCanAccess()`, then has its own create-perm branch. `groupAssignment` does a overlapping but not identical dance with `goto end` for repo collaborators.

This is the classic “working but messy” outcome the rubric rejects. There should be **one** `GroupAccess` evaluation that returns a typed capability struct:

```go
type GroupCapabilities struct {
    CanRead, CanWrite, CanAdmin, CanCreate, IsMember, IsOwner bool
    HiddenByAncestor bool
}
```

Web, API, and repo permission code should consume that, not each re-encode policy.

---

### Owner resolution is copy-pasted four times on `Group`

`IsOwnedBy`, `IsMemberOf`, `IsAdminOf`, and `CanCreateIn` each independently:

1. `GetUserByID(ownerID)`
2. Branch individual vs org
3. Call a different org helper

That is four nearly identical control flows in one model. Extract `loadOwnerContext(ctx) (individual | org, error)` and delete three copies.

---

### Hierarchy helpers use inconsistent strategies

- `GetParentGroupChain` — iterative queries (N round trips)
- `ChildGroupCond` — recursive CTE
- `AccessibleParentGroupCond` — recursive CTE + eager execution
- `ParentGroupCond` — iterative chain, logs on error, returns `NotIn`

Readers must understand three different mechanisms for one tree. Unify on CTE builders with bound parameters (you already have `groupHierarchyCTEBuilder` as a start).

---

### `ToAPIGroup` guarantees N+1 on list endpoints

```31:42:services/convert/repo_group.go
	apiGroup.NumSubgroups, err = group_model.CountGroups(...)
	_, apiGroup.NumRepos, err = repo_model.SearchRepositoryByCondition(...)
```

Every group in `GetOrgGroups`, `GetUserGroups`, `GetGroupSubGroups` triggers 2 extra queries. For a tree UI this becomes brutal quickly. Either:

- batch-count in the list handler, or
- make counts optional/lazy in API responses.

---

## 3. Spaghetti / branching growth

### `GroupAssignmentWeb` uses `goto` for collaborator escape hatch

```222:228:services/context/group.go
			if (opts.RequireMember && !repoGroup.IsMember) ||
				(opts.RequireOwner && !repoGroup.IsOwner) {
				if ctx.Repo.Repository != nil && ctx.Repo.Permission.HasAnyUnitAccess() {
					goto end
				}
				ctx.NotFound(nil)
```

This special case (“not a group member, but repo collaborator”) is exactly the kind of bolt-on branch that should live in the unified access model, not as a `goto` in template-prep middleware.

### `createCommonGroup` can fail silently

```52:63:routers/api/v1/group/group.go
	if err := group_service.NewGroup(ctx, group, ctx.Doer); err != nil {
		if group_model.IsErrGroupTooDeep(err) { ... }
		else if org_model.IsErrOrgNotExist(err) { ... }
		// ...
		return nil  // no ctx.APIError for unmatched errors
	}
```

Unmatched errors return `nil` without writing a response. That is a footgun and a sign the handler is doing too much error taxonomy inline.

### Magic number `20` instead of `NestingLimit`

```70:74:services/group/group.go
		if len(gidChain) >= 20 {
			return group_model.ErrGroupTooDeep{ ID: g.ParentGroupID }
		}
```

The model defines `NestingLimit = 20`. The service hardcodes `20`. Small, but symptomatic of boundaries not being owned cleanly.

---

## 4. Boundary / abstraction problems

### Two different path namespaces for the same repo

| Concern | Format |
|---------|--------|
| Disk / git storage | `owner/{groupID}/repo.git` |
| Web/API URLs | `owner/group/{groupID}/repo` |

This may be intentional, but the branch spreads both forms across ~80 Go files without a single canonical “repo locator” type. Every callsite rebuilds segments manually (`groupSegmentWithTrailingSlash`, `RelativePath`, `fillPathParts`, env vars, swagger, etc.).

**Prefer:** one `RepoLocator { Owner, GroupID, Name }` with `StoragePath()`, `WebPath()`, `APIPath()`, `ParseFromHTTPPath()`. Right now path knowledge leaks everywhere.

### `models/shared/group/org_group.go` is a pass-through

24 lines wrapping `SearchRepository` with an accessible-ID filter. This does not earn its package boundary; it belongs in `models/group` or `models/repo` next to existing list helpers.

### `RepoGroup.UnitPermission` assumes org team semantics

```53:67:services/context/group.go
	teams, err := organization.GetUserOrgTeams(ctx, g.Group.OwnerID, doer.ID)
	// ...
	return teams.UnitMaxAccess(unitType)
```

User-owned groups fall through to public visibility only. If that is intended, document it in the capability model. If not, this is feature logic applied with the wrong abstraction.

---

## 5. File size / decomposition

| File | Lines | Concern |
|------|------:|---------|
| `routers/web/web.go` | 1831 | Already monolithic; branch adds mirrored routes |
| `models/group/group.go` | 612 | Hierarchy + ACL + mutation + SQL generation |
| `routers/api/v1/group/group.go` | 551 | Fine for an API surface, but handler error handling is repetitive |
| `services/context/group.go` | 320 | Assignment + template data + API wrapper |

`models/group/group.go` should split before it crosses 1k:

- `hierarchy.go` — CTEs, chains, move, cycle
- `access.go` — conditions, `CanAccess*`
- `crud.go` — find/get/update

Do this **before** more subgroup features land.

---

## 6. What is actually good

Credit where due:

- `MoveGroup` / `CheckCycle` extraction and unit tests in `models/group/group_test.go` are focused and test real ordering behavior.
- `MoveGroupItem` unifying repo + group moves behind one service entry is the right direction (delete should reuse it).
- Integration coverage in `tests/integration/api_repo_group_test.go` is substantive — team permutations, nested visibility, collaborator edge cases.
- Route handler extraction (`rootRepoFn`, etc.) shows awareness of DRY; it just stops one level too early.

---

## Approval bar checklist

| Criterion | Status |
|-----------|--------|
| No structural regression | **Fail** — route duplication, permission sprawl |
| No obvious code-judo missed | **Fail** — normalize paths/routes; unify hierarchy ACL |
| No unjustified file-size explosion | **Fail** — `web.go` already huge, got bigger |
| No spaghetti branching | **Fail** — `goto`, scattered group checks |
| No magical/indirect abstractions | **Fail** — DB-in-Cond builder |
| Canonical layer reuse | **Fail** — delete bypasses move/rename primitive |
| Atomic orchestration | **Fail** — DB-only reparent on delete |

GamerGirlandCo committed:
fix: build errors after rebase

GamerGirlandCo committed:
chore: regenerate swagger templates

GamerGirlandCo committed:
refactor: clean up repo route registration

GamerGirlandCo committed:
refactor: update `models/group/group.go`

- replace `AccessibleParentGroupCond` with inline CTE in `IsPrivateBecauseOfParentPermissions`
- add tests for `IsPrivateBecauseOfParentPermissions`

GamerGirlandCo committed:
refactor: update group model package

- break out access/permission-related methods inside `group.go` into `access.go`
- add `Capabilities` struct and associated `GetCapabilities` computation method

GamerGirlandCo committed:
refactor: update `services/context` and `api/v1` package

consume and check capabilities returned from `Group.GetCapabilities()` instead of checking perms in multiple places

GamerGirlandCo committed:
refactor: update `models/group/access.go`

remove repeated code across `IsAdminOf`, `IsMemberOf` and `IsOwnedBy` methods

GamerGirlandCo committed:
refactor: update group model

update all funcs that create CTEs to use new internal func, `groupHierarchyCTEBuilder`

GamerGirlandCo committed:
fix: ensure that all errors are handled inside `createCommonGroup`

GamerGirlandCo committed:
refactor: update group service

use `NestingLimit` constant instead of hardcoding `20`

GamerGirlandCo committed:
style: run formatter

GamerGirlandCo committed:
refactor: break up `models/group/group.go` further

GamerGirlandCo committed:
style: run formatter

GamerGirlandCo committed:
fix: panic during API group assignment

GamerGirlandCo committed:
fix: nil dereference in `reqGroupMembership`

GamerGirlandCo committed:
fix: update `models/group/access.go`

fix potential bug where group ownership is required but not granted while doer has admin access

GamerGirlandCo committed:
style: remove unused parameter from `groupAssignment`

GamerGirlandCo committed:
fix: syntax error when testing on mysql

GamerGirlandCo committed:
fix: error when testing on mssql

`Cannot perform an aggregate function on an expression containing an aggregate or a subquery.`

GamerGirlandCo committed:
refactor: update `services/context/group.go`

create `AddGroupBreadcrumbs` func to add breadcrumb data and helpers to web context

GamerGirlandCo committed:
fix: ensure that repositories are moved on disk when deleting groups non-recursively

GamerGirlandCo committed:
refactor: update how group paths for urls/storage are made

create new `giturl.Locator` struct with common methods, replace repeated group segment construction with calls to those methods

GamerGirlandCo committed:
style: fix lint errors

GamerGirlandCo committed:
fix: update `UnitPermission` method

- don't assume org/team semantics unless group owner is an org, otherwise fall through to deriving permissions from the `Capabilities` struct

GamerGirlandCo committed:
fix: build errors after rebase

bircni commented:
> Code review starts when you open a non-draft PR or move a draft out of draft state. After that, do not rebase or squash your branch; it makes new changes harder to review.

https://github.com/go-gitea/gitea/blob/main/CONTRIBUTING.md#maintaining-open-prs

aol-nnov commented:
A bit of off topic, but @bircni I'm impressed by https://github.com/go-gitea/gitea/pull/35265#issuecomment-4679656141

What model was that? How many tokens did that review cost?

bircni commented:
> A bit of off topic, but @bircni I'm impressed by [#35265 (comment)](https://github.com/go-gitea/gitea/pull/35265#issuecomment-4679656141)
> 
> What model was that? How many tokens did that review cost?

this was composer 2.5 fast with the [thermo nuclear review skill](https://github.com/cursor/plugins/blob/main/cursor-team-kit/skills/thermo-nuclear-code-quality-review/SKILL.md)
also I pasted it some findings I found but yes I was also impressed

Cannot say much about the cost but it was less than 1% of monthly usage with pro

GamerGirlandCo committed:
chore: regenerate swagger definitions

GamerGirlandCo commented:
@bircni icymi, i've implemented the changes you (and claude) have requested. what should i do about the merge conflicts, since i can't rebase? 😅

wxiaoguang commented:
I think I have said clearly in #1872: document your design first. Clarify various edge cases first.

For example: How to handle "group" permissions? What the URL path will be? How to make it work with Actions (which only supports "owner/repo" layout)?

Without a clear design, I believe there are many things wrong in code.

wxiaoguang commented:
> what should i do about the merge conflicts, since i can't rebase? 😅

1. Rebase or not, it doesn't matter now. I don't think anyone has the time or ability to review the 117 commits in this PR, too many things are already mixed together
2. If you'd like to have a clear commit history and make some progresses, extract some changes into small PRs and make them merged first, especially the framework-level functions, the changes (refactoring) which also benefit current code base. These small PRs can be merged soon and you can learn more about the code base by participating in these PRs

wxiaoguang commented:
And, a kind suggestion: stay away from AI slop.


Just take a look at the first file in your commits: What does "//nolint:depguard // this package wraps it" mean? It was only used in our "json" package, our "json" package wraps the Golang's stdlib json package. But you or your AI just copied it blindly. That's just the first file in this `+57,222` PR, so, I have no confidence for what the other code would be.


<img width="2130" height="1080" alt="image" src="https://github.com/user-attachments/assets/13bd34fd-2f99-47ee-bee7-38a307fae064" />


<img width="1282" height="508" alt="image" src="https://github.com/user-attachments/assets/cd7c1435-f221-4454-a339-0e5596515cfe" />

codedbygo commented:
I went through this whole thread plus #35265 before writing anything, since a few of you have asked for years now for a design to actually land here before more code gets written. Figured I'd take a shot at it.

Quick recap of why this keeps stalling, because I think it matters for picking an approach: back in 2017 bkcsoft already called out that full nested groups need a large rewrite of the org model and won't be backwards compatible. Still true today — Organization in the codebase is literally type Organization user_model.User, orgs are rows in the same user table as people, and they share one global unique username. There's no parent field anywhere. lafriks floated a much lighter "virtual subgroup" idea back in 2020 that never got built, and wxiaoguang said in 2024 this was "nearly impossible... too many blockers." Then #35265 came along this year with a real implementation attempt, and got stuck on exactly the three questions wxiaoguang keeps asking: how do group permissions work, what's the URL going to look like, and how does this not break Actions (which only knows owner/repo). There was also a smaller attempt (#37775) that just added a ParentTeamID column with no actual inheritance logic and got closed for being incomplete.

So I think the real problem isn't lack of effort, it's that nobody's picked one of three fairly different shapes this feature could take, and everyone's kind of assumed the GitLab-identical one by default without weighing what it costs.

Option 1: keep it purely cosmetic. This is basically what lafriks suggested in 2020. Add a grouping concept that's just for organizing repos visually in the UI — breadcrumbs, filtering — but the repo's actual owner, URL, and API identity don't change at all. Nothing about permissions, Actions, git clone URLs, webhooks, anything changes, because nothing about how a repo is addressed moves. Cheapest and safest by far, one nullable column and a join table. Downside is obvious: it's not real nesting, no separate permissions per subgroup, no delegated ownership of a subgroup. Some people asking for this feature specifically want that part.

Option 2: subgroups are just organizations with a parent. This is what lunny suggested in the #35265 thread and I think it's underrated. Give Organization a nullable ParentOrgID. A subgroup literally is an org, so it gets the entire existing team/permission/repo-owner system for free instead of us building a parallel one. Permission inheritance would mean walking up the parent chain (capped at a small depth, so it stays cheap) — if you're an owner of the parent, you're implicitly an owner of the children. Because a subgroup is still a row in the user table with a unique name, repos inside it still resolve as a normal flat /subgroupname/repo URL, so Actions, webhooks, LFS, git-http, existing API clients, none of that needs to change. That answers all three of wxiaoguang's questions basically by construction rather than needing special-case handling. The real cost, and I want to be upfront about it rather than bury it: subgroup names still eat into the single global username/org namespace, same as orgs already do today. So myorg/backend and someone else's org called backend still can't coexist. That's a real limitation, not a nice one, but it's the same limitation Gitea already lives with today for orgs vs users, just extended one level down.

Option 3: a genuinely new Group entity, which is roughly what #35265 currently does — separate from the user table, its own permission model, nested URL paths, new endpoints. Closest to actual GitLab parity, arbitrary depth, groups aren't users so no namespace collision. But it's also the most expensive by far: a whole new permission-check code path running alongside the existing org one (bircni already flagged duplication risk in review), a new URL resolution layer, and the Actions question doesn't have a clean answer yet under this option — would repos need a flattened owner/repo alias just for Actions while the UI uses the nested path? That needs to be settled before more code gets written, not after.

My take: go with option 2. It reuses code that already exists and is already tested instead of adding a second parallel system, it doesn't touch URLs or Actions at all so backward compatibility is basically free, and it can be split into small PRs that land independently the way wxiaoguang already asked for on #35265 — first just the schema and migration with no behavior change, then permission inheritance as its own testable piece, then the UI (creation flow, breadcrumbs, settings), then exposing parent/child on the existing org API instead of standing up a whole new group API.

Also I'd scope v1 down on purpose: cap depth at maybe 2-3 levels instead of GitLab's 20 (that number always felt arbitrary anyway), and skip subgroups under personal accounts for now — that got argued both ways in the #35265 thread and I don't think it needs to be solved in the same breath as the org case.

Couple things I'm genuinely unsure about and would want opinions on before anyone writes more code:

Is the shared-namespace cost in option 2 actually acceptable, or is it a hard blocker that pushes this toward option 3 despite the extra cost?
Is a small depth cap for v1 fine, or does this need to support arbitrary depth from day one?
Given lunny already nudged toward this same direction on #35265, does it make more sense to work this into that PR with GamerGirlandCo rather than starting something separate? I'd rather coordinate than end up with two competing implementations.
Not trying to steamroll the existing PR here, just trying to get an actual decision made so the next attempt at this doesn't stall the same way the last three did.

ARNik commented:
For many users, Gitea subgroups are needed for migration from GitLab. Therefore, it is essential to accurately reproduce the nested URL structure. If subgroups are only cosmetic, it defeats the purpose.

wxiaoguang commented:
> If subgroups are only cosmetic, it defeats the purpose.

Yes, that's the root reason that why I insist on that there should be a complete design first, but not just hacky patch the code base.

codedbygo commented:
> > If subgroups are only cosmetic, it defeats the purpose.
> 
> Yes, that's the root reason that why I insist on that there should be a complete design first, but not just hacky patch the code base.

Agreed, and that's basically why I walked it back from cosmetic in the last reply — a subgroup that's just a label isn't the feature people are actually asking for. The design on the table now (real nested URL for browsing/clone, stable flat identity underneath for Actions/API) is meant to be a complete answer to that, not a patch. If any piece of it still looks hand-wavy to you, tell me which part and I'll go deeper before anyone touches code.

codedbygo commented:
> For many users, Gitea subgroups are needed for migration from GitLab. Therefore, it is essential to accurately reproduce the nested URL structure. If subgroups are only cosmetic, it defeats the purpose.

Fair point, and it's not just option 1 — option 2 doesn't give you a nested URL either, just better permissions underneath a flat path. If you're migrating from gitlab.com/company/backend/api, a flat /backend-api/repo really doesn't cut it.

Here's the catch: a fully nested URL is the exact thing #35265 has been stuck on for months, because Gitea's router today is strictly owner/repo, and Actions only understands that shape too. So "just nest it" isn't free.

What I'd propose instead: let the browser URL and git clone path be genuinely nested (/group/subgroup/project), resolved through a new lookup — but keep a stable flat identifier underneath for Actions/webhooks/API, same idea as how a repo rename gets redirected rather than breaking every integration. That gets you GitLab-shaped URLs for humans without reopening the Actions problem.

GamerGirlandCo commented:
> > For many users, Gitea subgroups are needed for migration from GitLab. Therefore, it is essential to accurately reproduce the nested URL structure. If subgroups are only cosmetic, it defeats the purpose.
> 
> Fair point, and it's not just option 1 — option 2 doesn't give you a nested URL either, just better permissions underneath a flat path. If you're migrating from gitlab.com/company/backend/api, a flat /backend-api/repo really doesn't cut it.
> 
> Here's the catch: a fully nested URL is the exact thing #35265 has been stuck on for months, because Gitea's router today is strictly owner/repo, and Actions only understands that shape too. So "just nest it" isn't free.
> 
> What I'd propose instead: let the browser URL and git clone path be genuinely nested (/group/subgroup/project), resolved through a new lookup — but keep a stable flat identifier underneath for Actions/webhooks/API, same idea as how a repo rename gets redirected rather than breaking every integration. That gets you GitLab-shaped URLs for humans without reopening the Actions problem.

or we could just fork `go-chi/chi` lol

mrmelon54 commented:
> > For many users, Gitea subgroups are needed for migration from GitLab. Therefore, it is essential to accurately reproduce the nested URL structure. If subgroups are only cosmetic, it defeats the purpose.
> 
> Fair point, and it's not just option 1 — option 2 doesn't give you a nested URL either, just better permissions underneath a flat path. If you're migrating from gitlab.com/company/backend/api, a flat /backend-api/repo really doesn't cut it.
> 
> Here's the catch: a fully nested URL is the exact thing #35265 has been stuck on for months, because Gitea's router today is strictly owner/repo, and Actions only understands that shape too. So "just nest it" isn't free.
> 
> What I'd propose instead: let the browser URL and git clone path be genuinely nested (/group/subgroup/project), resolved through a new lookup — but keep a stable flat identifier underneath for Actions/webhooks/API, same idea as how a repo rename gets redirected rather than breaking every integration. That gets you GitLab-shaped URLs for humans without reopening the Actions problem.

My first thought while reading option 2 was about the username conflicts. What if the flat username for a group contains the parent usernames using an underscore or other separator which is allowed by actions and the API. Then both the nested and flat URLs could be supported for the web interface, cloning, and anything else. Except for 

* Flat: `/<org>-<group>-<group2>/<repo>`
* Nested: `/<org>/<group>/<group2>/<repo>`

The initial work for this approach could add a new type to the users table specifically for the nested groups, which would be useful for a future rework to the permission system to support nested access controls.

The nesting could default to 5 layers as a nice middle group between no nesting and the 20 layers supported by GitLab. If future work is done to allow nested permission controls, then a caching layer seems appropriate to retain the speed using lookups.

It could be useful for the API to support using IDs directly for repositories:
* `/repos/{id}/avatar`
* `/repos/{id}/branches`

codedbygo commented:
> > > For many users, Gitea subgroups are needed for migration from GitLab. Therefore, it is essential to accurately reproduce the nested URL structure. If subgroups are only cosmetic, it defeats the purpose.
> > 
> > 
> > Fair point, and it's not just option 1 — option 2 doesn't give you a nested URL either, just better permissions underneath a flat path. If you're migrating from gitlab.com/company/backend/api, a flat /backend-api/repo really doesn't cut it.
> > Here's the catch: a fully nested URL is the exact thing #35265 has been stuck on for months, because Gitea's router today is strictly owner/repo, and Actions only understands that shape too. So "just nest it" isn't free.
> > What I'd propose instead: let the browser URL and git clone path be genuinely nested (/group/subgroup/project), resolved through a new lookup — but keep a stable flat identifier underneath for Actions/webhooks/API, same idea as how a repo rename gets redirected rather than breaking every integration. That gets you GitLab-shaped URLs for humans without reopening the Actions problem.
> 
> My first thought while reading option 2 was about the username conflicts. What if the flat username for a group contains the parent usernames using an underscore or other separator which is allowed by actions and the API. Then both the nested and flat URLs could be supported for the web interface, cloning, and anything else. Except for
> 
> * Flat: `/<org>-<group>-<group2>/<repo>`
> * Nested: `/<org>/<group>/<group2>/<repo>`
> 
> The initial work for this approach could add a new type to the users table specifically for the nested groups, which would be useful for a future rework to the permission system to support nested access controls.
> 
> The nesting could default to 5 layers as a nice middle group between no nesting and the 20 layers supported by GitLab. If future work is done to allow nested permission controls, then a caching layer seems appropriate to retain the speed using lookups.
> 
> It could be useful for the API to support using IDs directly for repositories:
> 
> * `/repos/{id}/avatar`
> * `/repos/{id}/branches`

Good refinement. One tweak: use a doubled separator, --, not single. Gitea's username validation already forbids consecutive separator chars, so no existing name can contain -- — that makes the flattened name unambiguous and reversible, not just unique (a single - would let foo-bar the org collide with foo + subgroup bar).

Dedicated user.Type for groups is the right call over my ParentOrgID-implies-group approach — cleaner, though it does mean auditing existing Type == UserTypeOrganization checks across the codebase, worth saying upfront.

+1 on the ID-based endpoints, and I'd land that independently regardless of subgroups — GitHub does the same with /repositories/{id}. Depth 5 sounds right. Agreed on caching mattering eventually, though I'd check what permission-caching already exists before assuming we need new infrastructure.

codedbygo commented:
> > > For many users, Gitea subgroups are needed for migration from GitLab. Therefore, it is essential to accurately reproduce the nested URL structure. If subgroups are only cosmetic, it defeats the purpose.
> > 
> > 
> > Fair point, and it's not just option 1 — option 2 doesn't give you a nested URL either, just better permissions underneath a flat path. If you're migrating from gitlab.com/company/backend/api, a flat /backend-api/repo really doesn't cut it.
> > Here's the catch: a fully nested URL is the exact thing #35265 has been stuck on for months, because Gitea's router today is strictly owner/repo, and Actions only understands that shape too. So "just nest it" isn't free.
> > What I'd propose instead: let the browser URL and git clone path be genuinely nested (/group/subgroup/project), resolved through a new lookup — but keep a stable flat identifier underneath for Actions/webhooks/API, same idea as how a repo rename gets redirected rather than breaking every integration. That gets you GitLab-shaped URLs for humans without reopening the Actions problem.
> 
> or we could just fork `go-chi/chi` lol

Ha, no need to fork chi — it already does wildcard catch-alls elsewhere in this repo (Maven package download uses r.Get("/*", ...)). Same trick works here: one wildcard route that swallows the whole path, then a resolver in the handler walks segments against known group/repo names to figure out where the group chain ends and the repo begins. That's roughly how GitLab and Docker registries handle arbitrary-depth paths too — router stays dumb, resolution logic lives in application code.

mrmelon54 commented:
> Good refinement. One tweak: use a doubled separator, --, not single. Gitea's username validation already forbids consecutive separator chars, so no existing name can contain -- — that makes the flattened name unambiguous and reversible, not just unique (a single - would let foo-bar the org collide with foo + subgroup bar).
> 
> Dedicated user.Type for groups is the right call over my ParentOrgID-implies-group approach — cleaner, though it does mean auditing existing Type == UserTypeOrganization checks across the codebase, worth saying upfront.
> 
> +1 on the ID-based endpoints, and I'd land that independently regardless of subgroups — GitHub does the same with /repositories/{id}. Depth 5 sounds right. Agreed on caching mattering eventually, though I'd check what permission-caching already exists before assuming we need new infrastructure.

All good ideas.

I also forgot to add to my original comment that the group depth should be configurable to allow for increasing it to the 20 limit supported by GitLab, just in case someone relies on that. But the default makes more sense being 5.
